ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/target/VirtualArtifactRepository.java

```diff
@@ -0,0 +1,207 @@
+				null);
+		Assert.isTrue(targetLocation.isResolved());
+		for (TargetBundle targetBundle : targetLocation.getBundles()) {
+			if (!targetBundle.getStatus().isOK()) {
+				PDECore.log(Status.warning(NLS.bind(Messages.VirtualArtifactRepository_0, targetBundle)));
+				continue;
+			}
+			BundleInfo bundleInfo = targetBundle.getBundleInfo();
+			IArtifactKey artifactKey = BundlesAction.createBundleArtifactKey(bundleInfo.getSymbolicName(),
+					bundleInfo.getVersion());
+			IArtifactDescriptor artifactDesriptor = new ArtifactDescriptor(artifactKey);
+			artifacts.put(artifactDesriptor, bundleInfo);
+		}
+		for (TargetFeature targetFeature : targetLocation.getFeatures()) {
+			IArtifactKey artifactKey = FeaturesAction.createFeatureArtifactKey(targetFeature.getId(),
+					targetFeature.getVersion());
+			IArtifactDescriptor artifactDesriptor = new ArtifactDescriptor(artifactKey);
+			artifacts.put(artifactDesriptor, targetFeature.getFeatureModel());
+		}
+	}
+
+	private static String getLocationSafe(ITargetLocation targetLocation) {
+		try {
+			return targetLocation.getLocation(false);
+		} catch (CoreException e) {
+			return "<unknown>"; //$NON-NLS-1$
+		}
+	}
+
+	@Override
+	public IStatus getRawArtifact(IArtifactDescriptor descriptor, OutputStream destination, IProgressMonitor monitor) {
+		Object artifactModel = artifacts.get(descriptor);
+		if (artifactModel == null) {
+			return Status.error(NLS.bind(Messages.VirtualArtifactRepository_1, descriptor));
+		}
+		try {
+			transferArtifact(artifactModel, destination);
+			return Status.OK_STATUS;
+		} catch (Exception e) {
+			return Status.error(e.getLocalizedMessage(), e);
+		}
+	}
+
+	private void transferArtifact(Object artifactModel, OutputStream destination) throws Exception {
+		if (artifactModel instanceof BundleInfo bundleInfo) {
+			URI location = bundleInfo.getLocation();
+			if (location == null) {
+				throw new FileNotFoundException(bundleInfo.getSymbolicName());
+			}
+			try (InputStream is = location.toURL().openStream()) {
+				is.transferTo(destination);
+			}
+		} else if (artifactModel instanceof IFeatureModel featureModel) {
+			String installLocation = featureModel.getInstallLocation();
+			if (installLocation != null) {
+				File featureJar = new File(installLocation);
+				if (featureJar.isFile()) {
+					Files.copy(featureJar.toPath(), destination);
+					return;
+				}
+			}
+			IFeature feature = featureModel.getFeature();
+			// Generate in-memory feature jar (with only the feature.xml)
+			Manifest manifest = new Manifest();
+			manifest.getMainAttributes().put(Attributes.Name.MANIFEST_VERSION, "1.0"); //$NON-NLS-1$
+
+			JarOutputStream jos = new JarOutputStream(destination, manifest);
+			jos.putNextEntry(new JarEntry("feature.xml")); //$NON-NLS-1$
+			PrintWriter printWriter = new PrintWriter(jos);
+			feature.write("", printWriter); //$NON-NLS-1$
+			printWriter.flush();
+			jos.closeEntry();
+
+			jos.flush();
```

Comment:
Looking at the implementation of `flush()` I got the impression that it is not necessary to call `jos.flush()`.
Furthermore maybe calling `jos.finish()` will also implicitly close the only entry and thus `jos.closeEntry()` might be not necessary as well.  At least `putNextEntry()` implicitly closes a previously opened entry.
But I cannot tell with confidence and would need to be checked.

Reply:
The call to `closeEntry()` is indeed redundant. In hindsight, I would also expect the stream to write any buffered data, when `finish()` is called. So I assume that's also not strictly necessary.

```
public void finish() throws IOException {
	ensureOpen();
	if (finished) {
		return;
	}
	if (current != null) {
		closeEntry();
	}
	// write central directory
	long off = written;
	for (XEntry xentry : xentries)
		writeCEN(xentry);
	writeEND(off, written - off);
	finished = true;
}
```

Reply:
> The call to `closeEntry()` is indeed redundant. In hindsight, I would also expect the stream to write any buffered data, when `finish()` is called. So I assume `flush()` also not strictly necessary.

Yes, especially since the destination does not have to be flushed (at least here, if it is necessary it is the responsibility of the caller).